xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -361,15 +365,143 @@ static FilterChain parseFilterChain(
     if (name.isEmpty()) {
       name = UUID.randomUUID().toString();
     }
+    FilterChainMatch filterChainMatch = parseFilterChainMatch(proto.getFilterChainMatch());
+    checkForUniqueness(uniqueSet, filterChainMatch);
     return new FilterChain(
         name,
-        parseFilterChainMatch(proto.getFilterChainMatch()),
+        filterChainMatch,
         httpConnectionManager,
         downstreamTlsContext,
         tlsContextManager
     );
   }
 
+  private static void checkForUniqueness(Set<FilterChainMatch> uniqueSet,
+      FilterChainMatch filterChainMatch) throws ResourceInvalidException {
+    if (uniqueSet != null) {
+      Set<FilterChainMatch> crossProduct = getCrossProduct(filterChainMatch);
```

Comment:
This is debatably incorrect behavior. If an entry contains a duplicate within itself, the gRFC's documented behavior of converting to disjunctive normal form would NACK it. I hadn't considered that case when making the gRFC, but I've confirmed that Envoy would NACK it as well. So this would be invalid:

```
prefix_ranges: 192.168.0.0/16, 192.168.0.0/16
```

Now I would agree that the behavior presented here does not disagree with "If the most-specific matching logic might not produce a unique result, then the Listener must be NACKed", as the result is still unique. But I'd rather not have an unnecessary subtle behavior difference as well as pay an additional cost for it. The code here can return a list from `getCrossProduct()` and use `if (!uniqueSet.add(cur))` below.

Reply:
Interesting. When I was writing the code I intentionally put `uniqueSet.addAll(crossProduct);` at the end to avoid checking duplicates within a `FilerChain` since the intent was really to ensure a unique FilterChain and not to detect useless repetition of the same value in a repeated field.

But I will make this change and also add a test to detect that condition which basically rejects FilterChainMatch's that contain a useless but harmless repetition of a value.

Reply:
FWIW, the addAll() at the end does nothing because the product is returned as a Set.

Reply:
> the addAll() at the end does nothing 

The `uniqueSet` is maintained across all the filterChains of a listener and because of this `addAll` I am able to find duplicate matches across all those filterChains.

Reply:
Fixed this via `checkAndAdd` method.

Reply:
I was trying to communicate that if you used `if (!uniqueSet.add(cur))` instead of contains() then you wouldn't need addAll. Doing the add _after_ checking all the individual matchers in the product was unnecessary. (Or, using a Set was unnecessary; you could have either done contains() or addAll(), but you didn't need both.)

Reply:
Good point. I'll make that change

Reply:
> Fixed this via checkAndAdd method.

Why not just swap to a List as I suggested? I don't understand the benefit of the checkAndAdd approach.

Reply:
checkAndAdd seemed fine to me for a moment, as I thought you were passing `uniqueSet` to `getCrossProduct()`. That would be a fine approach. I really expect there to only be one collection where we are checking for duplicates; I don't see much point to other options.

Reply:
Oh, because each expand creates a new Set, you have to repeat checkAndAdd a bunch. Still, seems swapping to a List is strictly easier and less error-prone.

Reply:
Done.